# КОММЕНТАРИЙ РЕВЬЮВЕРА
Привет Максим 👋!)

Меня зовут Сибирев Владислав, я осуществлял проверку проекта. Может перейдем сразу на <b>ты</b>? Если это неприемлемо, то обязательно дай знать)

Далее по ходу работы ты сможешь встретить мои комментарии, они будут отмечены разными цветами:

<div class="alert alert-block alert-success">
 ✔️ Похвала и возможности развития
</div>
<div class="alert alert-block alert-warning">
⚠️ Некритичные замечания, но стоит обратить внимание
</div>
<div class="alert alert-block alert-danger">
❌ Грубая ошибка, необходимо исправить
</div>


Заранее прошу не удалять мои комментарии, это важно. Если у тебя возникнут вопросы, обозначь их, я с радостью тебе помогу!)

Как вариант, можешь обозначить их так:
<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [4]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [5]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [8]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [9]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [11]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [13]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [14]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [15]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [16]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [17]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [18]:
data.duplicated().sum()

71

In [19]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
children_set = data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count() * 100
children_set.sort_values()

children
5    0.000000
0    7.543822
3    8.181818
1    9.234609
2    9.454191
4    9.756098
Name: debt, dtype: float64

In [26]:
grouped_children_set = data.groupby('children')['debt'].count()
grouped_children_set

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [27]:
# children_pivot = data.pivot_table(index = ['children'], \
#                                 columns = ['debt'], \
#                                values = 'purpose', aggfunc='count')
# children_pivot['ratio'] = children_pivot[1] / children_pivot[0] * 100
# children_pivot.sort_values(by='ratio')

In [1]:
children_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
children_pivot

NameError: name 'data' is not defined

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ По выводу таблицы все отлично, только мы `ratio` мы получаем не от количество людей кто хорошо отдает, а от общего количества кредиторов)) Можно оптимизировать код следующим образом, но это на твое усмотрение:
    
    
       data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
   
<br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь все отлично стало)) Все замечания исправил)) 

  </div>

**Вывод:** Есть небольшая зависимость между количеством детей и возвратом кредитов в срок. 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Не бойся использовать в выводах цифры, также нужно сделать акцент на сбалансированности выборок. Примерно выводы будут выглядеть так: 
`люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка`)) <br>

<div class="alert alert-block alert-info">
<b>Полагаю нет смысла править вывод к этому пункту, так как вы сами его написали. Спасибо за пример.</b>
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Хорошо, договорились)

  </div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [32]:
family_status_set = data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count() * 100
family_status_set.sort_values()

family_status
вдовец / вдова           6.624606
в разводе                7.064760
женат / замужем          7.560558
гражданский брак         9.313014
Не женат / не замужем    9.763948
Name: debt, dtype: float64

In [33]:
#family_status_pivot = data.pivot_table (index = ['family_status'], \
 #                                      columns = ['debt'], \
 #                                      values = 'family_status_id', aggfunc='count')
#family_status_pivot['ratio'] = family_status_pivot[1] / family_status_pivot[0] * 100
#family_status_pivot.sort_values(by='ratio')

In [45]:
family_status_pivot = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
family_status_pivot.sort_values(by='mean')

count      mean    sum
family_status                                  
вдовец / вдова           951.0  0.066246   63.0
в разводе               1189.0  0.070648   84.0
женат / замужем        12261.0  0.075606  927.0
гражданский брак        4134.0  0.093130  385.0
Не женат / не замужем   2796.0  0.097639  273.0

**Вывод:** Лучше всего отдают кредит вдовы/вдовцы, люди в разводе и женатые/замужние, доля невозврата составляет 6.6% , 7.0% и 7.5% соответственно. Хуже отдают кредиты должники, находящиеся в гражданском браке и неженатые/незамужние клиенты, доля невозврата составляет 9.3% и 9.7% соответственно.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Выводы должны быть детальными) В выводах давай будет использовать цифры, также  удели здесь внимание тоже сбалансированности выборок))

<div class="alert alert-block alert-info">
<b>Здесь я так понимаю, не стоит пропускать выборку по вдовам/вдовцам и людям в разводе. Все-таки их не так мало и эти данные имеют роль. Поправьте, пожалуйста, если я не прав.</b>
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  В любом случае надо указать, что эта выборка малочислена))

  </div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [53]:

total_income_set = data.groupby('total_income_category')['debt'].sum() / \
data.groupby('total_income_category')['debt'].count() * 100
total_income_set.sort_values(ascending=False)

total_income_category
E    9.090909
C    8.498210
A    8.000000
B    7.060231
D    6.017192
Name: debt, dtype: float64

In [54]:
#total_income_pivot = data.pivot_table(index='total_income_category', \
#                                     columns='debt', \
#                                     values='income_type', aggfunc='count')
#total_income_pivot['ratio'] = total_income_pivot[1] / total_income_pivot[0] * 100
#total_income_pivot.sort_values(by='ratio', ascending=False)

In [55]:
total_income_pivot = data.pivot_table(index = ['total_income_category'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
total_income_pivot.sort_values(by='count', ascending=False)

count      mean     sum
total_income_category                           
C                      15921.0  0.084982  1353.0
B                       5014.0  0.070602   354.0
D                        349.0  0.060172    21.0
A                         25.0  0.080000     2.0
E                         22.0  0.090909     2.0

**Вывод:** Есть зависимость от уровня доходов. Люди, чей доход составляет 50001-200000(категория С) хуже отдают кредиты в срок, здесь доля невозврата составляет 8.4%. Лучше обстоят дела с клиентами, чей доход составляет 200001–1000000(категория B), доля должников составляет 7.0%. Остальные категории сравнивать нецелесообразно, так как выборка по ним незначительна.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь сравниваем только две группы B и С, не забываем про сбалансированость выборки))<br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь сравнил две самые большие группы))

  </div>

In [56]:
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [57]:
purpose_set = data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count() * 100
purpose_set.sort_values()

purpose_category
операции с недвижимостью    7.255139
проведение свадьбы          7.911803
получение образования       9.252758
операции с автомобилем      9.347978
Name: debt, dtype: float64

In [58]:
#purpose_pivot = data.pivot_table(index='purpose_category', \
 #                                columns='debt', \
 #                                values='purpose', aggfunc='count')
#purpose_pivot['ratio'] = purpose_pivot[1] / purpose_pivot[0] * 100
#purpose_pivot.sort_values(by='ratio')

In [59]:
purpose_pivot = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
purpose_pivot.sort_values(by='count', ascending=False)

count      mean    sum
purpose_category                                  
операции с недвижимостью  10751.0  0.072551  780.0
операции с автомобилем     4279.0  0.093480  400.0
получение образования      3988.0  0.092528  369.0
проведение свадьбы         2313.0  0.079118  183.0

**Вывод:** Люди, которые берут кредиты на операции с автомобилем и на получение образования, хуже отдают кредиты в срок, доля невозврата составляет 9.3% и 9.2% соответственно. Лучше отдают кредиты клиенты, взявшие кредит на проведение свадьбы, процент невовзрата составляет 7.9. Лучше остальных кредиты возвращают люди, взявшие ипотеку, здесь процент составляет 7.2%. Более вероятно, что люди, которые берут ипотеку более серьезно подходят к этому решению. Как правило, чтобы взять ипотеку нужен первый взнос, который в среднем составляет от 10% от стоимости жилья/помещения, и который выплачивается из собственных средств заемщика. Это подразумевает, что человек, скорее всего, долго подходил к решению о покупке жилья и в большей мере рассчитал свои возможности по выплатам.  

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ Почему по твоему самые надежные заемщики те, кто берет ипотеку? Здесь можно порасуждать в выводах)) 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Самые большие суммы при ипотеке, поэтому банки предъявляют самые жесткие требования к ипотечникам, поэтому их проверяют вдоль и поперек при выдаче кредитов, плюс собственность находится в залоге у банка, не будешь платить она вернется к банку)) 
   
       
  </div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Технологические ошибки(сбой при копировании, выгрузке). Человеческий фактор(опечатки). В нашем случае пропуски в столбце 'days_employed' и 'total_income' одинаковы, и могут быть связаны с тем, что эти люди никогда не работали, так как молоды или работали не официально, либо данные некорректно выгружены. 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️  Да, всё здраво. В принципе эти возможные причины сразу и приходят в голову. 
    
Однако, если пристально взглянуть на на наши данные, то увидим что пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество ячеек с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Маловероятно, что специалисты по выдаче кредитов, заполнив все без исключения ячейки других столбцов по каждому кредитополучателю, оставляли незаполненными именно два показателя, и сделали это ровно 2174 раза в каждом столбце. Даже не проводя проверку, во всех ли случаях незаполненные ячейки находятся в одной и то же строке с высокой долей вероятности можно утверждать, что произошла техническая ошибка при формировании таблицы с исходными данными (либо некий сбой во время выгрузки данных на анализ, либо при объединении данных в единый файл, который также должен был передаваться на анализ, либо эти данные не собирались).
</div>

<div class="alert alert-block alert-info">
<b>Комментарий усвоил. Спасибо за объяснение.</b>
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение лучше характеризует данные, если некоторые значения сильно выделяются среди большинства.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Медиана не подвержена выбросам, все верно)) 

### Шаг 4: общий вывод.

- Присутствует зависимость между количеством детей и возвратом кредита в срок. Люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%.
- Наблюдается зависимость между семейным положением и возвратом кредита в срок. Лучше всего отдают кредит вдовы/вдовцы, люди в разводе и женатые/замужние, доля невозврата составляет 6.6% , 7.0% и 7.5% соответственно. Хуже отдают кредиты должники, находящиеся в гражданском браке и неженатые/незамужние клиенты, доля невозврата составляет 9.3% и 9.7% соответственно.
- Есть зависимость от уровня доходов. Люди, чей доход составляет 50001-200000(категория С) хуже отдают кредиты в срок, здесь доля невозврата составляет 8.4%. Лучше обстоят дела с клиентами, чей доход составляет 200001–1000000(категория B), доля должников составляет 7.0%. Остальные категории сравнивать нецелесообразно, так как выборка по ним незначительна.
- Люди, которые берут кредиты на операции с автомобилем и на получение образования, хуже отдают кредиты в срок, доля невозврата составляет 9.3% и 9.2% соответственно. Лучше отдают кредиты клиенты, взявшие кредит на проведение свадьбы, процент невовзрата составляет 7.9. Лучше остальных кредиты возвращают люди, взявшие ипотеку, здесь процент составляет 7.2%

# Рекомендации для банка
- Уделить внимание правилам валидации на сервере, чтобы больше не сталкиваться с проблемой дублей.
- Считать стаж работы в годах, пользователи чаще всего не знают свой стаж в днях.
- Показывать операторам в банке или пользователям на сайте подсказки заранее, когда вводят сверхвысокие/сверхнизкие
  значения в полях: возраст, стаж, количество детей.

 <div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Выводы отлично переработал, молодец)) Рекомендации даны банку, исследование проведено))
  
  
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера:</b> 

❌ Выводы должны быть детальными:
- Отвечаешь на вопрос "Есть ли зависимость между количеством детей и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между семейным положением и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Как разные цели кредита влияют на его возврат в срок?"
- Даешь рекомендацией банку по улучшению сбора данных и скоринга.

Общий вывод - резюме этой работы. Большинство заказчиков посмотрят сначала его (а некоторые только его). Чтобы выводы не огорошивали заказчика своей очевидностью, стоит добавить в них цифр, всяких корреляций, процентов итд :) Сделай так, чтобы вся найденная в данных информация была отражена в выводе))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>
    
Максим, отлично поработал, видно что умеешь работать с кодом, в общем молодец, но нужно будет оптимизировать код при изучении взаимосвязей и поработать над выводами)) 
<br> Жду исправленного варианта)
  </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>
    
Максим, отлично поработал, можно двигаться дальше)) Успехов в дальнейшем обучении)) 
  </div>